### PR TITLE
api v1 - render errors with rabl

### PR DESCRIPTION
**Pull request by pitr-ch**
better detection of permission failure in model

### CODE REVIEW

File: app/models/authorization.rb [hunk 1]
```diff
@@ -30,7 +30,8 @@ def enforce_permissions operation
     klasses = klass.pluralize
     return true if User.current and User.current.allowed_to?("#{operation}_#{klasses}".to_sym)
 
-    errors.add :base, "You do not have permission to #{operation} this #{klass}"
+    errors.add :base, message = "You do not have permission to #{operation} this #{klass}"
+    errors.add :permission, message
```
Comment (pitr-ch) [inline app/models/authorization.rb]: what do you thing about this one @ohadlevy ?
I did not want to remove the base error, there can be some dependencies on it. But I wanted to make it clear what type of error it is. 
see https://github.com/theforeman/foreman/pull/95/files#L0R19

Comment (pitr-ch) [inline app/models/authorization.rb]: never-mind, I've rewritten this, see https://github.com/theforeman/foreman/pull/95/files#diff-1

File: app/controllers/api/base_controller.rb [hunk 2]
```diff
@@ -6,37 +6,42 @@ class BaseController < ActionController::Base
...
-                 end
-               end
-
-      # set 403 status on permission errors
-      if errors.any? { |error| error =~ /You do not have permission/ }
-        options[:json_code] = :forbidden
-      end
+    rescue_from StandardError, :with => lambda { |error|
+      Rails.logger.error "#{error.message} (#{error.class})\n#{error.backtrace.join("\n")}"
+      render :template => '/api/v1/errors/standard_error', :status => 500, :locals => { :error => error }
```
Comment (ohadlevy) [inline app/controllers/api/base_controller.rb]: looks like we need a helper method to return the api version etc

File: app/controllers/api/base_controller.rb [hunk 3]
```diff
@@ -6,37 +6,42 @@ class BaseController < ActionController::Base
...
 
-      render :json => { "errors" => errors }, :status => options[:json_code]
+    def resource_class
+      @resource_class ||= resource_name.camelize.constantize
     end
 
+    protected
+
+    def process_resource_error(options = { })
+      resource = options[:resource] ||= get_resource || raise("No error to process")
```
Comment (ohadlevy) [inline app/controllers/api/base_controller.rb]: wouldnt get resource raise before the last raise in this line ?

**Comment by ohadlevy** [comment]
one thing that bothers me, is that we skip the controller permissions checks, and directly go to the model permissions checks.

foreman currently have a way to map permissions into web actions, and this basically means, that those permissions are only valid for the UI part of the site, and maybe we should consider mapping the api actions as well.

File: app/controllers/api/base_controller.rb (hunk 3 above)
Comment (pitr-ch) [inline app/controllers/api/base_controller.rb]: there should be: 

``` ruby
resource = options[:resource] || get_resource || raise("No resource to process")
```

I'll fix it in next commit

**Comment by pitr-ch** [comment]
I've fixed the permission skiping in https://github.com/theforeman/foreman/pull/92/files, see #authorize. I think the pull-92 is implementing what you are describing.

File: app/controllers/api/base_controller.rb (hunk 2 above)
Comment (pitr-ch) [inline app/controllers/api/base_controller.rb]: fixed in baf2a47d41ef62e475c1db96242e3b4881d62a8b

**Comment by pitr-ch** [comment]
> I'm not 100% sure, but if get_resource raise by it self, I dont think you'll ever get to the raise here.

You are right, fixed.

> nice catch, did you try to make sure it works? we currently have no tests for this code :)

I added some test for manager rights, it seems to be working ok :)

**Comment by ohadlevy** [comment]
i mean the version matcher, not the permissions

**Comment by pitr-ch** [comment]
>  out of the routes?

``` ruby
Rails.application.routes.routes.select { |r| r.verb == 'GET' && r.path =~ /\/api\// && r.path !~ /:id/ }
```

but it is not perfect, i look into it next week 

> rebase

yeah, it's not i'll do the squash and rebase

**Comment by ohadlevy** [comment]
ok, please rebase :)

**Comment by ohadlevy** [comment]
Please dont forget about https://github.com/theforeman/foreman/pull/95#issuecomment-6727357

**Pull request merged**